File: floe/jobdeck/plan.py
"""Deck -> placement plan + report: the M1 deliverable.

`plan_deck` ties parser, catalog, geometry and colour together without
touching the renderer. Its output (placements with mag/dx/dy on the deck
grid, a colour per placement, the skip ledger) is exactly what the M2
composite scene will consume, so the numbers are gated here first.
"""

from __future__ import annotations

import json
import os
from dataclasses import asdict

from .color import ColorScheme, MODE_IDENTIFIER, normalize_mode, order_text
from .geom import MISSING_RAISE, plan
from .parser import parse_jobdeck
from .sources import SourceCatalog


def plan_deck(deck_path: str, sources_dir=None, ids=None,
              mode: str = MODE_IDENTIFIER, missing: str = MISSING_RAISE,
              scheme: ColorScheme | None = None, cross: bool = True,
              safety: int = 2, strict: bool = True, load_ids=None):
    """Parse, probe, place and colour one deck.

    Returns (deck, catalog, placements, stats, scheme, colormap).
    `sources_dir` defaults to the deck's directory (TC paths are
    relative to it). `ids` is the ANALYSIS selection (`floe2 jobdeck
    --level`): it restricts the placements and, in chip mode, splices
    the selected levels into the colour order (section 2). `load_ids`
    is the LOAD selection (the viewer, `floe2 view/index/render
    --level`): it restricts the placements, the rows and the sources
    probed in full, and moves NO colour - review 2026-09-10 (9th)
    P2-4: a CHIP loaded with one level kept its full-deck colour
    only in the level view. The grid never moves with either: the
    whole deck's extent and every source's dbu (header only for the
    sources a load selection leaves out) decide it.
    """
    mode = normalize_mode(mode)
    deck = parse_jobdeck(deck_path, strict=strict)
    if sources_dir is None:
        sources_dir = os.path.dirname(os.path.abspath(deck_path)) or "."
    catalog = SourceCatalog(sources_dir)
    if load_ids is not None:
        load_ids = sorted({int(i) for i in load_ids}) or None
    catalog.probe_all(deck.sources(load_ids))
    dbus = dict(catalog.dbus())
    if load_ids is not None:
        # the grid contract: every source's dbu, from the header alone
        # for the sources the load leaves out (no cache-state lookup,
        # not registered as probed)
        for tc in deck.sources():
            if tc not in catalog.infos:
                dbu = catalog.header_dbu(tc)
                if dbu:
                    dbus[tc] = dbu
    if scheme is None:
        scheme = ColorScheme(mode=mode, cross_ly_dt=cross)
    else:
        scheme.mode = mode
    by_chip = scheme.mode == "chip"
    placements, stats = plan(deck, dbus, safety=safety,
                             cross=scheme.cross_ly_dt,
                             ids=load_ids if load_ids is not None else ids,
                             missing=missing, by_chip=by_chip,
                             bad=catalog.bad())
    colormap = scheme.build(deck, ids)
    stats["colors"] = {
        "mode": scheme.mode,
        "order": scheme.order_table(deck, ids),
        "map": {_ckey(k): v for k, v in colormap.items()},
    }
    stats["sources"] = catalog.report()
    return deck, catalog, placements, stats, scheme, colormap


def _ckey(k) -> str:
    if isinstance(k, tuple):
        return "%d/%d" % k
    return str(k)


def report_dict(deck, placements, stats) -> dict:
    """JSON-ready report: the deck's own report, the plan statistics, and
    every placement (mag/dx/dy in um and on the deck grid)."""
    st = dict(stats)
    st.pop("out_of", None)
    return {
        "deck": deck.report(),
        "plan": st,
        "placements": [asdict(p) for p in placements],
    }


def write_report(path: str, deck, placements, stats) -> None:
    with open(path, "w") as fh:
        json.dump(report_dict(deck, placements, stats), fh, indent=1,
                  default=_json_default)


def _json_default(o):
    if isinstance(o, (set, frozenset)):
        return sorted(o)
    if isinstance(o, tuple):
        return list(o)
    return str(o)


def deck_summary(deck, catalog, placements, stats, scheme) -> list[str]:
    """Human-readable summary lines (the `floe2 jobdeck` output)."""
    src = stats["sources"]
    cov = deck.coverage()
    lines = [
        "deck      : %s%s" % (deck.path,
                              "  (%s)" % deck.jb_name if deck.jb_name else ""),
        "chips     : %d  levels %s  (%d complete, %d partial)"
        % (len(deck.chips), deck.levels(), cov["complete"],
           cov["partial"]),
        "sources   : %d probed, %d ok, %d indexed (.floe)  dir %s"
        % (src["probed"], src["ok"], src["indexed"], catalog.dir),
        "instances : %d placed%s" % (
            stats["instances"],
            "" if stats["selection"] is None else
            " of %d (selection %s)" % (stats["instances_total"],
                                        stats["selection"])),
        "grid      : dbu %g um (%s)" % (stats["dbu"], stats["dbu_choice"]["why"]),
        "residual  : max %g um, %d non-zero, %d over half a dbu"
        % (stats["residual_max_um"], stats["residual_nonzero"],
           stats["residual_over_half_dbu"]),
        "mag       : %s" % stats["mags"],
        "bbox um   : %s" % (
            "none" if stats["bbox_um"] is None else
            "%.4f %.4f %.4f %.4f" % tuple(stats["bbox_um"])),
        "view      : %s -> %s" % (
            {"level": "level view (by mask level)",
             "chip": "chip view (by CHIP block)",
             "layer": "source layer view (LY/DT)"}.get(scheme.mode,
                                                        scheme.mode),
            order_text(stats["colors"]["order"])),
    ]
    for info in src["files"]:
        if info["status"] != "ok":
            lines.append("source    : %s %s (%s)" % (
                info["tc"], info["status"].upper(), info["error"]))
    for rec in stats["skipped"]:
        lines.append("skipped   : CHIP %s $%d %s: %s (%s, %d row(s))" % (
            rec["chip"], rec["idx"], rec["tc"], rec["reason"],
            rec["detail"], rec["rows"]))
    for rec in stats["deck_issues"]:
        lines.append("outside   : CHIP %s $%d %s: %s (not selected)" % (
            rec["chip"], rec["idx"], rec["tc"], rec["reason"]))
    for ln, msg in deck.warnings:
        lines.append("warning   : line %d: %s" % (ln, msg))
    ex = deck.extras()
    if ex:
        lines.append("unknown $ : %s" % ", ".join(
            "%s (%d)" % (k, len(v)) for k, v in sorted(ex.items())))
    if deck.unknown:
        lines.append("unparsed  : %d line(s), first: %d: %s" % (
            len(deck.unknown), deck.unknown[0][0], deck.unknown[0][1]))
    return lines


def placement_lines(placements, scheme, colormap) -> list[str]:
    out = ["%-8s %-4s %-4s %-24s %-8s %-9s %16s %16s %8s" % (
        "chip", "idx", "row", "tc", "ly/dt", "mag", "dx_um", "dy_um",
        "colour")]
    for p in placements:
        out.append("%-8s $%-3d %-4d %-24s %-8s %-9g %16.4f %16.4f %8s" % (
            p.chip, p.idx, p.row, p.tc, "%d/%d" % (p.ly, p.dt), p.mag,
            p.dx_um, p.dy_um, scheme.color_of(colormap, p)))
    return out

File: rust/app-core/src/jobdeck/plan.rs
//! Shared analysis service. Analysis selections and load selections are
//! deliberately different: only the former splice historical CHIP colors.
use super::{
    color::{ColorMap, ColorScheme, Mode},
    geom::{self, PlanOptions, PlannedDeck},
    index::validate_levels,
    parser::{general, JobDeck},
    sources::SourceCatalog,
    view::ViewRows,
};
use crate::{artifact, cache, check_cancelled, Result};
use serde_json::{json, Value};
use std::collections::BTreeSet;
use std::path::{Path, PathBuf};
use std::sync::atomic::AtomicUsize;

#[derive(Clone, Debug)]
pub struct AnalysisOptions {
    pub sources: Option<PathBuf>,
    pub ids: Option<BTreeSet<i64>>,
    pub load_ids: Option<BTreeSet<i64>>,
    pub mode: Mode,
    pub scheme: Option<ColorScheme>,
    pub cross: bool,
    pub skip_missing: bool,
    pub strict: bool,
}
impl Default for AnalysisOptions {
    fn default() -> Self {
        Self {
            sources: None,
            ids: None,
            load_ids: None,
            mode: Mode::Level,
            scheme: None,
            cross: true,
            skip_missing: false,
            strict: true,
        }
    }
}
#[derive(Debug)]
pub struct Analysis {
    pub deck: JobDeck,
    pub catalog: SourceCatalog,
    pub model: PlannedDeck,
    pub scheme: ColorScheme,
    pub colors: ColorMap,
}
impl Analysis {
    pub fn open(path: &Path, options: &AnalysisOptions, cancelled: &AtomicUsize) -> Result<Self> {
        check_cancelled(cancelled)?;
        let deck = JobDeck::read(path, options.strict, cancelled)?;
        let path = cache::absolute(path)?;
        let mut catalog = SourceCatalog::new(
            options
                .sources
                .as_deref()
                .unwrap_or(path.parent().expect("absolute path")),
        )?;
        // Empty LOAD means all, as in the GTK load dialog. Empty ANALYSIS
        // selection means no placements and must not be silently widened.
        let load = options.load_ids.as_ref().filter(|ids| !ids.is_empty());
        validate_levels(&deck, load)?;
        catalog.probe_all(deck.sources(load), cancelled)?;
        let mut dbus = catalog.dbus();
        if load.is_some() {
            for tc in deck.sources(None) {
                if !catalog.infos.contains_key(tc) {
                    if let Some(dbu) = catalog.header_dbu(tc, cancelled)? {
                        dbus.insert(tc.into(), dbu);
                    }
                }
            }
        }
        let mut scheme = options.scheme.clone().unwrap_or_else(|| ColorScheme {
            mode: options.mode,
            cross_ly_dt: options.cross,
            ..Default::default()
        });
        scheme.mode = options.mode;
        scheme.validate()?;
        let model = geom::plan(
            &deck,
            &dbus,
            &catalog.bad(),
            &PlanOptions {
                cross: scheme.cross_ly_dt,
                by_chip: scheme.mode == Mode::Chip,
                selected: load.or(options.ids.as_ref()).cloned(),
                skip_missing: options.skip_missing,
                ..Default::default()
            },
            cancelled,
        )?;
        let colors = scheme.build(&deck, options.ids.as_ref())?;
        check_cancelled(cancelled)?;
        Ok(Self {
            deck,
            catalog,
            model,
            scheme,
            colors,
        })
    }
    pub fn view_rows(&self, cancelled: &AtomicUsize) -> Result<ViewRows> {
        ViewRows::build(
            &self.deck,
            &self.model.stats,
            &self.scheme,
            &self.colors,
            cancelled,
        )
    }
    pub fn report(&self) -> Result<Value> {
        let mut stats = serde_json::to_value(&self.model.stats).expect("finite plan statistics");
        stats["colors"] = self.colors.report(self.scheme.mode);
        stats["sources"] = self.catalog.report();
        Ok(json!({"deck":self.deck.report()?,"plan":stats,"placements":self.model.placements}))
    }
    /// Local export preflight, including unselected/missing source names.
    /// A report must never replace the deck, an OASIS or a cache/lock through
    /// an alias. No read lease/hot-reload guarantee is implied here.
    pub fn output_path(&self, path: &Path, extra_inputs: &[PathBuf]) -> Result<PathBuf> {
        let mut files = vec![PathBuf::from(&self.deck.path)];
        files.extend_from_slice(extra_inputs);
        let mut trees = Vec::new();
        for tc in self.deck.sources(None) {
            let source = self.catalog.resolve(tc);
            let directory = cache::cache_path(&source)?;
            let mut lock = directory.as_os_str().to_owned();
            lock.push(".index.lock");
            files.push(source);
            files.push(lock.into());
            trees.push(directory);
        }
        artifact::protected_output(path, &files, &trees)
    }
    pub fn summary(&self) -> Result<Vec<String>> {
        let stats = &self.model.stats;
        let src = self.catalog.report();
        let coverage = self.deck.coverage()?;
        let levels: Vec<_> = self.deck.levels().into_iter().collect();
        let selection = stats
            .selection
            .as_ref()
            .map(|s| s.iter().copied().collect::<Vec<_>>());
        let mut lines = vec![
            format!(
                "deck      : {}{}",
                self.deck.path,
                if self.deck.name().is_empty() {
                    String::new()
                } else {
                    format!("  ({})", self.deck.name())
                }
            ),
            format!(
                "chips     : {}  levels {:?}  ({} complete, {} partial)",
                self.deck.chips.len(),
                levels,
                coverage["complete"],
                coverage["partial"]
            ),
            format!(
                "sources   : {} probed, {} ok, {} indexed (.floe)  dir {}",
                src["probed"],
                src["ok"],
                src["indexed"],
                self.catalog.directory.display()
            ),
            format!(
                "instances : {} placed{}",
                stats.instances,
                selection.map_or_else(String::new, |s| format!(
                    " of {} (selection {s:?})",
                    stats.instances_total
                ))
            ),
            format!(
                "grid      : dbu {} um ({})",
                general(stats.dbu),
                stats.dbu_choice.why
            ),
            format!(
                "residual  : max {} um, {} non-zero, {} over half a dbu",
                general(stats.residual_max_um),
                stats.residual_nonzero,
                stats.residual_over_half_dbu
            ),
            format!("mag       : {:?}", stats.mags),
            format!(
                "bbox um   : {}",
                stats.bbox_um.map_or_else(
                    || "none".into(),
                    |b| format!("{:.4} {:.4} {:.4} {:.4}", b[0], b[1], b[2], b[3])
                )
            ),
            format!(
                "view      : {} -> {}",
                match self.scheme.mode {
                    Mode::Level => "level view (by mask level)",
                    Mode::Chip => "chip view (by CHIP block)",
                    Mode::Layer => "source layer view (LY/DT)",
                },
                self.colors
                    .order
                    .iter()
                    .map(|r| format!(
                        "{} {}{}",
                        r.key,
                        r.name,
                        if r.source == "pinned" {
                            " (pinned)"
                        } else {
                            ""
                        }
                    ))
                    .collect::<Vec<_>>()
                    .join(", ")
            ),
        ];
        for i in self.catalog.infos.values().filter(|i| i.status != "ok") {
            lines.push(format!(
                "source    : {} {} ({})",
                i.tc,
                i.status.to_ascii_uppercase(),
                i.error
            ));
        }
        for r in &stats.skipped {
            lines.push(format!(
                "skipped   : CHIP {} ${} {}: {} ({}, {} row(s))",
                r.chip, r.idx, r.tc, r.reason, r.detail, r.rows
            ));
        }
        for r in &stats.deck_issues {
            lines.push(format!(
                "outside   : CHIP {} ${} {}: {} (not selected)",
                r.chip, r.idx, r.tc, r.reason
            ));
        }
        for (line, msg) in &self.deck.errors {
            lines.push(format!("error     : line {line}: {msg}"));
        }
        for (line, msg) in &self.deck.warnings {
            lines.push(format!("warning   : line {line}: {msg}"));
        }
        let report = self.deck.report()?;
        let extras = report["undocumented_entry_fields"]
            .as_object()
            .expect("extras object");
        if !extras.is_empty() {
            lines.push(format!(
                "unknown $ : {}",
                extras
                    .iter()
                    .map(|(k, v)| format!("{k} ({})", v.as_array().expect("extras rows").len()))
                    .collect::<Vec<_>>()
                    .join(", ")
            ));
        }
        if let Some((n, s)) = self.deck.unknown.first() {
            lines.push(format!(
                "unparsed  : {} line(s), first: {n}: {s}",
                self.deck.unknown.len()
            ));
        }
        Ok(lines)
    }
    pub fn placement_line(&self, p: &geom::Placement) -> String {
        format!(
            "{:<8} ${:<3} {:<4} {:<24} {:<8} {:<9} {:16.4} {:16.4} {:>8}",
            p.chip,
            p.idx,
            p.row,
            p.tc,
            format!("{}/{}", p.ly, p.dt),
            general(p.mag),
            p.dx_um,
            p.dy_um,
            self.scheme.color_of(&self.colors, p)
        )
    }
}
